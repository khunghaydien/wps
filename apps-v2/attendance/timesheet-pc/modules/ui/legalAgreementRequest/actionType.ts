import ROOT from '../actionType';

export default `${ROOT}/LEGAL-AGREEMENT-REQUEST` as const;
