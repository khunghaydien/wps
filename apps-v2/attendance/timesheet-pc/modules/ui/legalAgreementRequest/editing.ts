import {
  DISABLE_ACTION,
  DisableAction,
  EDIT_ACTION,
  EditAction,
  getPerformableDisableAction,
  getPerformableEditAction,
  LegalAgreementRequest,
} from '@attendance/domain/models/LegalAgreementRequest';
import { Code } from '@attendance/domain/models/LegalAgreementRequestType';

import ROOT from './actionType';

export type State = {
  id: string | null;
  requestType: Code | null;
  editAction: EditAction;
  disableAction: DisableAction;
  isEditing: boolean;
};

const initialState: State = {
  id: null,
  requestType: null,
  editAction: EDIT_ACTION.NONE,
  disableAction: DISABLE_ACTION.NONE,
  isEditing: false,
};

const ACTION_TYPE_ROOT = `${ROOT}/EDITING` as const;

const ACTION_TYPE = {
  INIT: `${ACTION_TYPE_ROOT}/INIT`,
  START: `${ACTION_TYPE_ROOT}/START`,
  CLEAR: `${ACTION_TYPE_ROOT}/CLEAR`,
} as const;

type Init = {
  type: typeof ACTION_TYPE.INIT;
  payload: {
    request: LegalAgreementRequest;
    isEditing: boolean;
  };
};

type Start = {
  type: typeof ACTION_TYPE.START;
};

type Clear = {
  type: typeof ACTION_TYPE.CLEAR;
};

type Action = Init | Start | Clear;

export const actions = {
  initialize: (request: LegalAgreementRequest, isEditing = false): Init => ({
    type: ACTION_TYPE.INIT,
    payload: {
      request,
      isEditing,
    },
  }),
  startEditing: (): Start => ({
    type: ACTION_TYPE.START,
  }),
  clear: (): Clear => ({
    type: ACTION_TYPE.CLEAR,
  }),
};

export default function reducer(
  state: State = initialState,
  action: Action
): State {
  switch (action.type) {
    case ACTION_TYPE.INIT: {
      const { request, isEditing } = action.payload;
      const editAction = getPerformableEditAction(request);
      const disableAction = getPerformableDisableAction(request);
      return {
        id: request.id,
        requestType: request.requestType,
        isEditing: editAction === EDIT_ACTION.CREATE || isEditing,
        editAction,
        disableAction,
      };
    }

    case ACTION_TYPE.START: {
      return {
        ...state,
        isEditing: true,
      };
    }

    case ACTION_TYPE.CLEAR:
      return { ...initialState };

    default:
      return state;
  }
}
