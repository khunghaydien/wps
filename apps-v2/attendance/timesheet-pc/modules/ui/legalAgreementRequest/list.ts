import { LegalAgreementRequestList } from '@attendance/domain/models/LegalAgreementRequest';

import ROOT from './actionType';

type State = LegalAgreementRequestList;

const initialState: State = {
  requests: [],
  availableRequestTypes: null,
};

const ACTION_TYPE_ROOT = `${ROOT}/LIST` as const;

const ACTION_TYPE = {
  SET: `${ACTION_TYPE_ROOT}/SET`,
  CLEAR: `${ACTION_TYPE_ROOT}/CLEAR`,
} as const;

type Set = {
  type: typeof ACTION_TYPE.SET;
  payload: LegalAgreementRequestList;
};

type Clear = {
  type: typeof ACTION_TYPE.CLEAR;
};

type Action = Set | Clear;

export const actions = {
  set: (list: LegalAgreementRequestList): Set => ({
    type: ACTION_TYPE.SET,
    payload: list,
  }),
  clear: (): Clear => ({
    type: ACTION_TYPE.CLEAR,
  }),
};

export default function reducer(
  state: State = initialState,
  action: Action
): State {
  switch (action.type) {
    case ACTION_TYPE.SET:
      return {
        ...(action as Set).payload,
      };

    case ACTION_TYPE.CLEAR:
      return {
        ...initialState,
      };

    default:
      return state;
  }
}
