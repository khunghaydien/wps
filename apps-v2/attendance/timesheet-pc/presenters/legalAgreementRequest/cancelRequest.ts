import { Store } from 'redux';

import {
  catchApiError,
  loadingEnd,
  loadingStart,
} from '@apps/commons/actions/app';

import { IPresenter } from '@attendance/application/useCaseInteractors/legalAgreementRequest/CancelRequestUseCaseInteractor';

export default ({ dispatch }: Store): IPresenter => ({
  start: () => {
    dispatch(loadingStart());
  },
  complete: () => {},
  error: (err) => {
    dispatch(catchApiError(err as Parameters<typeof catchApiError>[0]));
  },
  finally: () => {
    dispatch(loadingEnd());
  },
});
