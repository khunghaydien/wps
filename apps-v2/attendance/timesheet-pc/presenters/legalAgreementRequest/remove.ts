import { Store } from 'redux';

import {
  catchApiError,
  confirm,
  loadingEnd,
  loadingStart,
} from '@apps/commons/actions/app';
import msg from '@apps/commons/languages';

import { IPresenter } from '@attendance/application/useCaseInteractors/legalAgreementRequest/RemoveUseCaseInteractor';

export default ({ dispatch }: Store): IPresenter => ({
  confirmRemoving: async () => {
    dispatch(loadingEnd());
    const result = (await dispatch(
      // @ts-ignore
      confirm(msg().Appr_Msg_RequestConfirmRemove)
    )) as unknown as boolean;
    dispatch(loadingStart());
    return result;
  },
  start: () => {
    dispatch(loadingStart());
  },
  complete: () => {},
  error: (err) => {
    dispatch(catchApiError(err as Parameters<typeof catchApiError>[0]));
  },
  finally: () => {
    dispatch(loadingEnd());
  },
});
