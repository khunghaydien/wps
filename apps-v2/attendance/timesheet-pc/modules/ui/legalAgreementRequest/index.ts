import { combineReducers } from 'redux';

import editing from './editing';
import list from './list';
import page from './page';
import requests from './requests';

const reducers = {
  editing,
  requests,
  page,
  list,
};

const rootReducer = combineReducers(reducers);
export type State = ReturnType<typeof rootReducer>;

export default rootReducer;
