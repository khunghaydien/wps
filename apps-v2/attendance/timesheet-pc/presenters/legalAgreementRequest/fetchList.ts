import { Store } from 'redux';

import {
  catchApiError,
  loadingEnd,
  loadingStart,
} from '@apps/commons/actions/app';

import { State } from '@attendance/timesheet-pc/modules';
import { actions as listActions } from '@attendance/timesheet-pc/modules/ui/legalAgreementRequest/list';
import { actions as pageActions } from '@attendance/timesheet-pc/modules/ui/legalAgreementRequest/page';

import { IPresenter } from '@attendance/application/useCaseInteractors/legalAgreementRequest/FetchListUseCaseInteractor';

export default ({ dispatch, getState }: Store) =>
  (): IPresenter => {
    let opened = false;
    return {
      start: () => {
        const state = getState() as State;
        if (state.ui.legalAgreementRequest.page.opened) {
          opened = true;
          dispatch(loadingStart());
        }
        dispatch(pageActions.startLoading());
      },
      complete: (response) => {
        dispatch(listActions.set(response.requestList));
      },
      error: (err) => {
        dispatch(listActions.clear());
        dispatch(catchApiError(err as Parameters<typeof catchApiError>[0]));
      },
      finally: () => {
        if (opened) {
          dispatch(loadingEnd());
        }
        dispatch(pageActions.finishLoading());
      },
    };
  };
