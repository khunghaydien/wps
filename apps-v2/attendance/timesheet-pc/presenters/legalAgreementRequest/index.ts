import { Store } from 'redux';

import cancelApproval from './cancelApproval';
import cancelRequest from './cancelRequest';
import fetchList from './fetchList';
import reapply from './reapply';
import remove from './remove';
import submit from './submit';

export default (
  store: Store
): {
  fetchList: ReturnType<typeof fetchList>;
  submit: ReturnType<typeof submit>;
  cancelRequest: ReturnType<typeof cancelRequest>;
  cancelApproval: ReturnType<typeof cancelApproval>;
  remove: ReturnType<typeof remove>;
  reapply: ReturnType<typeof reapply>;
} => ({
  fetchList: fetchList(store),
  submit: submit(store),
  cancelRequest: cancelRequest(store),
  cancelApproval: cancelApproval(store),
  remove: remove(store),
  reapply: reapply(store),
});
