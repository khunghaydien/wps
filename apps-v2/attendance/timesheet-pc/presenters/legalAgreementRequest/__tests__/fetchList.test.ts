import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import cloneDeep from 'lodash/cloneDeep';

import presenter from '../fetchList';
import { IOutputData } from '@attendance/domain/useCases/legalAgreementRequest/IFetchListUseCase';

const mockStore = configureMockStore([thunk]);

beforeEach(() => {
  jest.clearAllMocks();
});

const dummyState = {
  ui: {
    legalAgreementRequest: {
      page: {
        opened: false,
      },
    },
  },
};

describe('complete()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore(dummyState);

    // Act
    presenter(store)().complete({
      targetDate: '2022-02-22',
      requestList: 'requestList',
    } as unknown as IOutputData);

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('error()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore(dummyState);

    // Act
    presenter(store)().error({
      message: 'Error Test',
    });

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('start()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore(dummyState);

    // Act
    presenter(store)().start();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('finally()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore(dummyState);

    // Act
    presenter(store)().finally();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('loading', () => {
  it('should use globalLoading', () => {
    // Arrange
    const state = cloneDeep(dummyState);
    state.ui.legalAgreementRequest.page.opened = true;
    const store = mockStore(state);
    const $presenter = presenter(store)();

    // Act
    $presenter.start();
    $presenter.finally();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should use only pageLoading', () => {
    // Arrange
    const state = cloneDeep(dummyState);
    state.ui.legalAgreementRequest.page.opened = false;
    const store = mockStore(state);
    const $presenter = presenter(store)();

    // Act
    $presenter.start();
    $presenter.finally();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});
