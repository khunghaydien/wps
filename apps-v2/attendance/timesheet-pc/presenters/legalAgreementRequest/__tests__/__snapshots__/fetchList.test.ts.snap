// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complete() should do. 1`] = `
Array [
  Object {
    "payload": "requestList",
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/LIST/SET",
  },
]
`;

exports[`error() should do. 1`] = `
Array [
  Object {
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/LIST/CLEAR",
  },
  Object {
    "payload": Object {
      "errorCode": undefined,
      "isContinuable": true,
      "isFunctionCantUseError": false,
      "message": "APIエラー - Error Test",
      "problem": "Error Test",
      "stackTrace": "",
      "type": "APIエラー",
    },
    "type": "CATCH_API_ERROR",
  },
]
`;

exports[`finally() should do. 1`] = `
Array [
  Object {
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/PAGE/FINISH_LOADING",
  },
]
`;

exports[`loading should use globalLoading 1`] = `
Array [
  Object {
    "type": "LOADING_START",
  },
  Object {
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/PAGE/START_LOADING",
  },
  Object {
    "type": "LOADING_END",
  },
  Object {
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/PAGE/FINISH_LOADING",
  },
]
`;

exports[`loading should use only pageLoading 1`] = `
Array [
  Object {
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/PAGE/START_LOADING",
  },
  Object {
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/PAGE/FINISH_LOADING",
  },
]
`;

exports[`start() should do. 1`] = `
Array [
  Object {
    "type": "TIMESHEET-PC/UI/LEGAL-AGREEMENT-REQUEST/PAGE/START_LOADING",
  },
]
`;
